### ISSUE TITLE
compatability with python 3.10

### ISSUE THREAD

**Issue by EMCP** [transferwise/pipelinewise-target-postgres#92]
I was utilizing this via meltano in a python 3.10 environment.. unfortunately.. it's failing and I think the key problem is a change in python itself

https://stackoverflow.com/questions/70943244/attributeerror-module-collections-has-no-attribute-mutablemapping

This workaround is as follows

```
import collections 
if sys.version_info.major == 3 and sys.version_info.minor >= 10

    from collections.abc import MutableMapping
else 
    from collections import MutableMapping
```

**Comment by cwe5590** [comment]
Looks like this is happening on this line: https://github.com/transferwise/pipelinewise-target-postgres/blob/5e8891d7a839c35dbc4e3d3e5dd021db483afac0/target_postgres/db_sync.py#L150

**Comment by EMCP** [comment]
bump, would love to see this patched.. so I can bump up to 22.04 .. thank you @cwe5590 for the help

**Comment by cwegener** [comment]
> 
> This workaround is as follows
> 
> ```
> import collections 
> if sys.version_info.major == 3 and sys.version_info.minor >= 10
> 
>     from collections.abc import MutableMapping
> else 
>     from collections import MutableMapping
> ```

I think this should be checking for Python 3.3 (**!!!**) or higher instead (thats when the deprecation was introduced ...)

https://docs.python.org/3.11/whatsnew/3.3.html#collections

**Comment by cwegener** [comment]
Ah. Looking at PR #97  - That's actually the correct way to do it. Since the CI only tests Python 3.6 and higher, why even bother trying to be compatible with Python 3.1 or Python 3.2 (I doubt that the tap would even run on those versions unmodified)

**Comment by Samira-El** [comment]
Hey, please use any Python from 3.7 to 3.9, this connector is not tested against 3.10

**Comment by cwegener** [comment]
> Hey, please use any Python from 3.7 to 3.9, this connector is not tested against 3.10

That logic does not make sense. Python 3.9 is not included in the Unit Test CI job either.

**Comment by sbelluzzo** [comment]
> > Hey, please use any Python from 3.7 to 3.9, this connector is not tested against 3.10
> 
> That logic does not make sense. Python 3.9 is not included in the Unit Test CI job either.

Agreed, though PR #97 also has changes to 

> Ah. Looking at PR #97 - That's actually the correct way to do it. Since the CI only tests Python 3.6 and higher, why even bother trying to be compatible with Python 3.1 or Python 3.2 (I doubt that the tap would even run on those versions unmodified)

That was my logic, especially considering that everything below Python 3.7 is [considered EOL](https://devguide.python.org/versions/) at this point anyway. If you're pinned to a lower version, you probably need to pin a lot of other dependencies too.

**Comment by sbelluzzo** [comment]
> > Hey, please use any Python from 3.7 to 3.9, this connector is not tested against 3.10
> 
> That logic does not make sense. Python 3.9 is not included in the Unit Test CI job either.

Agreed, though PR #97 included changes to include 3.9 and 3.10 in the CI test job. @Samira-El, I'm happy to break that into a separate PR (from the MutableMapping fix) to more clearly delineate the two changes, if needed.

**Comment by cwegener** [comment]
> > > Hey, please use any Python from 3.7 to 3.9, this connector is not tested against 3.10
> > 
> > 
> > That logic does not make sense. Python 3.9 is not included in the Unit Test CI job either.
> 
> It's being tested in https://github.com/transferwise/pipelinewise , though we must have forgotten to update the CI matrix in this connector.

Ah! Great!

@sbelluzzo 
So, I would suggest then to simply avoid referencing the term "Python 3.10" in this context of the `MutableMapping` updates, because technically the `MutableMapping` doesn't really have anything to do with Python 3.10. The alias `collections.MultableMapping` that provides legacy support for Python versions older than 3.3 should probably never have been relied upon in the first place in the `target-postgres`.

I would consider the removal of the `collections.*` alias in Python 3.10 to be rather co-incidental to this particular topic.

Does that make sense?

This way, Wise don't need to add Python 3.10 to their supported platforms (it's still pretty new at 1 year old :grin: )

And by removing the reliance on the `collections.*` alias that was introduced in Python 3.3, it will make the target work for those people who run it on Python 3.10

**Comment by EMCP** [comment]
Ubuntu 22.04 starts with python 3.10 .. and in setting up this in that
env... Breaks..   it's up to the project what they can support but.. this
is preventing me from utilizing in that env

**Comment by cwegener** [comment]
> Ubuntu 22.04 starts with python 3.10 .. and in setting up this in that env... Breaks.. it's up to the project what they can support but.. this is preventing me from utilizing in that env

Python 3.10 went GA 1 year ago and that's for how long I've been using it on my local development environment without issues. Rolling Release distros don't wait for an arbitrary amount of time to upgrade their packages :-)

Since @sbelluzzo  has already opened the PR, you can actually install from that PR in your pip_url like so:

```
   - name: target-postgres
     variant: transferwise
     pip_url: git+https://github.com/transferwise/pipelinewise-target-postgres/@refs/pull/97/head
```

**Comment by tayloramurphy** [comment]
@Samira-El is it possible to push an update to https://pypi.org/project/pipelinewise-target-postgres/ so that 3.10 is included when new installs are made?
